src/circuit_builder.rs

```diff
@@ -90,215 +443,138 @@ pub enum CircuitErrors {
+            let opening_key = circuit.opening_key().as_ref().cloned().unwrap();
+            let prover_key = circuit.prover_key().as_ref().cloned().unwrap();
+            let verifier_key = circuit.verifier_key().as_ref().cloned().unwrap();
+
             // Write the keys
             use std::fs::File;
             use std::io::Write;
+
+            let mut commit_file = File::create("ck_testcirc")?;
+            commit_file.write(commit_key.to_bytes()[..].as_ref())?;
+
+            let mut opening_file = File::create("ok_testcirc")?;
+            opening_file.write(opening_key.to_bytes()[..].as_ref())?;
+
             let mut prover_file = File::create("pk_testcirc")?;
             prover_file.write(prover_key.to_bytes()[..].as_ref())?;
+
             let mut verifier_file = File::create("vk_testcirc")?;
             verifier_file.write(verifier_key.to_bytes().as_ref())?;
         };
 
-        // Read ProverKey
-        let prover_key = ProverKey::from_bytes(&std::fs::read("pk_testcirc")?[..]).unwrap();
-        // Read VerifierKey
-        let verifier_key = VerifierKey::from_bytes(&std::fs::read("vk_testcirc")?[..]).unwrap();
+        let size = circuit.circuit_size();
 
-        // Generate new inputs
-        // Generate circuit compilation params
-        let inputs2 = [
-            BlsScalar::from(20u64),
-            BlsScalar::from(5u64),
+        // Read the keys
+        let ck = CommitKey::from_bytes(&std::fs::read("ck_testcirc")?[..]).unwrap();
+        let ok = OpeningKey::from_bytes(&std::fs::read("ok_testcirc")?[..]).unwrap();
+        let pk = ProverKey::from_bytes(&std::fs::read("pk_testcirc")?[..]).unwrap();
+        let vk = VerifierKey::from_bytes(&std::fs::read("vk_testcirc")?[..]).unwrap();
+
+        // Prove the statement
+        // Generate the proof inputs
+        let witness = TestCircuitData::witness_data(
             BlsScalar::from(25u64),
-            BlsScalar::from(100u64),
-        ];
-        // Replace previous used inputs by the new ones.
-        circuit.inputs = Some(&inputs2);
-
-        let public_inputs2 = vec![
-            PublicInput::BlsScalar(-BlsScalar::from(25u64), 0),
-            PublicInput::BlsScalar(-BlsScalar::from(100u64), 0),
-        ];
-        let proof = circuit.gen_proof(&pub_params, &prover_key, b"TestCirc")?;
-        circuit.verify_proof(
-            &pub_params,
-            &verifier_key,
-            b"TestCirc",
-            &proof,
-            &public_inputs2,
-        )
+            BlsScalar::from(5u64),
+            BlsScalar::from(30u64),
+            BlsScalar::from(125u64),
+        );
+        // Generate the proof
+        let proof = TestCircuit::from(witness)
+            .prepare_prover(size, pk, ck)
+            .gen_proof()?;
+
+        // Verify the proof
+        // Generate the public inputs
+        let public = TestCircuitData::public_data(BlsScalar::from(30u64), BlsScalar::from(125u64));
```

Comment:
The goal should be not to force the user to know the indexes. Indeed that's why I told above to roll back to my definition of the PI generator fn.

Reply:
He is not setting the indexes here. He passes the scalars as parameters, and `Circuit::gen_public_inputs` will set the index